### ISSUE TITLE
option: use numbered registers for smaller deletions

### ISSUE THREAD

**Issue by Himura2la** [neovim/neovim#8121]
Once upon a time, there was a [bug that was causing small deletions appear in the numbered registers](
https://groups.google.com/d/msg/vim_dev/pbN4rAw8LmI/jR54qQjvuOsJ) under some conditions. Then it was fixed.
And now there's no way to use such a useful bugfeature.
Ingo Karkat who reported this bug also proposed a super-useful option `smalldeletethreshold`, which I'm dying without.
I hope that Neovim development is not so sluggish as Vim development and this feature-request can survive. I even ready to try implementing it...

### Actual behaviour
Deletions like `d5w` has no history, though it's a kind of a large amount of text. If I use `d5w`, then jump to the insertion place and use `x` to fix a small thing, I'll lose the `d5w`'d text completely.

### Expected behaviour
I want to set a (non-existing) `smalldeletethreshold` option to 2 so that all deletions larger than `x` are pushed into the numbered registers "1.."9.

**Comment by justinmk** [comment]
I think vim would accept it if you implemented it with test coverage. We will too.

Personally I wish it worked for yanks, not just delete.

**Comment by Himura2la** [comment]
I'm afraid my ability for implementation is a little bit overestimated, I've even never done unit testing before...
And yes, of course, it should work for yanks. I just noticed that they call everything "deletions" when talking about registers

**Comment by Himura2la** [comment]
@justinmk Sorry, I was wrong about yanking. I'm new to vim and I did not know "1.."9 registers are for deletions only. I'm not sure it's good to change this, but it's definitely not the point of this issue. I think I've implemented the option in its most basic form and I did it global instead of buffer-specific for simplicity. Maybe I could make it buffer-specific

**Comment by panosl** [comment]
I made this with ChatGPT, which works with what I had in mind (saving each deletion into the numbered registry):

```
function! ShiftAndSetNumberedRegister()
  let temp_content = getreg('"')
  let temp_type = getregtype('"')

  for i in range(9, 1, -1)
    let from_reg = string(i-1)
    let to_reg = string(i)
    let old_content = getreg(from_reg)
    let old_type = getregtype(from_reg)
    call setreg(to_reg, old_content, old_type)
  endfor

  call setreg('1', temp_content, temp_type)
endfunction

augroup custom_delete_behavior
  autocmd!
  autocmd TextDelete * call ShiftAndSetNumberedRegister()
  autocmd TextChanged * call ShiftAndSetNumberedRegister()
augroup end
```

**Comment by MyyPo** [comment]
There was a [relevant issue](https://github.com/vim/vim/issues/7756) in the upstream, where it was preliminarily decided against adding a new option for the sake of backwards-compatibility opting instead for adding a new `"1-` motion for accessing small deletes and `g"1` for yanks.

I am not sure that it is really an unwanted breaking change, since it is opt-in and I guess that most plugins dealing with the current behavior are actually the ones trying to mimic similiar options, but also I see that adding new options is bad for testing and stability in general.

Either way there does not seem to be much progress in this direction, although, both features were requested several times on different platforms.

If anyone is looking for some simple fix the vimscript solution posted earlier works indeed:
```lua
vim.g.lat_small_reg = ""
vim.api.nvim_create_autocmd("TextYankPost", {
	callback = function()
		local small_val = vim.fn.getreg("-")
		if small_val ~= vim.g.lat_small_reg then
			local small_typ = vim.fn.getregtype("-")

			local val_old = vim.fn.getreg(1)
			local typ_old = vim.fn.getreg(1)

			vim.fn.setreg(1, small_val, small_typ)

			for i = 2, 9 do
				local copy_val = vim.fn.getreg(i)
				local copy_typ = vim.fn.getregtype(i)
				vim.fn.setreg(i, val_old, typ_old)
				val_old = copy_val
				typ_old = copy_typ
			end
		end
	end,
})

```

**Comment by asmodeus812** [comment]
Yeah second that, i think an option is probably a better approach, having yet another way to delete/yank text into registers to me is too confusing. A treshold option is great, since it allows you to fine tune what you consider a small delete, and by default it will be unset, which will default to the current behavior.

**Comment by justinmk** [comment]
`:help yankring` shows a small but effective example of how to use `TextYankPost` to change what is stored in registers 1-9. I don't see why this needs to be a builtin option.
